### PR TITLE
lock parent folders for the owner when locking a shared file as recipient

### LINKED ISSUE #17181: Properly lock share owner parent folders
### Steps
1. Enable transactional locking
2. Create users "user1" and "user2"
3. Login as "user1"
4. Create a folder "sub/sub2/folder"
5. Share "sub/sub2/folder" with "user2"
6. Login as "user2"
7. Create a file in "/folder/test.txt" with WebDAV (use slow upload by adding sleep(30) here https://github.com/owncloud/core/blob/master/lib/private/connector/sabre/file.php#L204)
8. During the upload, login as "user1" and rename "sub"

### Expected result
"sub" is locked

### Actual result
Rename succeeds.

### Versions
ownCloud 8.1 git master (58439c337c2e242c199f153d0529e0b0b482f03a)

Concern raised by @icewind1991 

When the parent folders of "test.txt" are selected to be locked in the owner's tree, they are only selected up to the parent root, which is "sub/sub2/folder". The correct behavior would be to also lock "sub/sub2" and "sub".

@icewind1991 said he was working on a POC and possible fix. The idea would be that whenever a shared storage root is being locked, the storage will automatically lock the parents in the original storage.

CC @jnfrmarks @SergioBertolinSG FYI about this test case

### PR DESCRIPTION
**Pull request by icewind1991**
Fixes #17181

cc @DeepDiver1975 @PVince81

### CODE REVIEW

**Comment by PVince81** [comment]
also @schiesbn @nickvergessen 

I'll test this soon

File: apps/files_sharing/lib/sharedmount.php [hunk 1]
```diff
@@ -41,34 +41,40 @@ class SharedMount extends MountPoint implements MoveableMount {
...
+	 */
+	private $recipientView;
+
 	public function __construct($storage, $mountpoint, $arguments = null, $loader = null) {
 		// first update the mount point before creating the parent
 		$this->ownerPropagator = $arguments['propagator'];
+		$this->recipientView = new View('/' . $arguments['user'] . '/files');
 		$newMountPoint = $this->verifyMountPoint($arguments['share'], $arguments['user']);
 		$absMountPoint = '/' . $arguments['user'] . '/files' . $newMountPoint;
+		$arguments['ownerView'] = new View('/' . $arguments['share']['uid_owner'] . '/files');
```
Comment (PVince81) [inline apps/files_sharing/lib/sharedmount.php]: How about instantiating ownerView directly in SharedStorage ?
The uid_owner is already available there.

(unless you wanted to be able to inject it for testing)

File: apps/files_sharing/lib/sharedstorage.php [hunk 2]
```diff
@@ -623,6 +629,10 @@ public function acquireLock($path, $type, ILockingProvider $provider) {
 		/** @var \OCP\Files\Storage $targetStorage */
 		list($targetStorage, $targetInternalPath) = $this->resolvePath($path);
 		$targetStorage->acquireLock($targetInternalPath, $type, $provider);
+		if ($path === '') {
```
Comment (PVince81) [inline apps/files_sharing/lib/sharedstorage.php]: Add a comment: "When locking the root of the target storage, also need to lock its parent folders on the owner's storage"

File: apps/files_sharing/tests/locking.php [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+	private $ownerUid;
+	private $recipientUid;
+
+	public function setUp() {
+		parent::setUp();
+
+		$lockingProvider = \OC::$server->getLockingProvider();
+		if ($lockingProvider instanceof MemcacheLockingProvider) {
+			$lockingProvider->releaseAll();
```
Comment (PVince81) [inline apps/files_sharing/tests/locking.php]: we already have this in TestCase, see https://github.com/owncloud/core/blob/master/tests/lib/testcase.php#L46

**Comment by PVince81** [comment]
Tested, works.

Please have a look at my comments.

File: apps/files_sharing/lib/sharedmount.php [hunk 4]
```diff
@@ -41,34 +41,40 @@ class SharedMount extends MountPoint implements MoveableMount {
...
+	/**
+	 * @var \OC\Files\View
+	 */
+	private $recipientView;
+
 	public function __construct($storage, $mountpoint, $arguments = null, $loader = null) {
 		// first update the mount point before creating the parent
 		$this->ownerPropagator = $arguments['propagator'];
+		$this->recipientView = new View('/' . $arguments['user'] . '/files');
 		$newMountPoint = $this->verifyMountPoint($arguments['share'], $arguments['user']);
```
Comment (nickvergessen) [inline apps/files_sharing/lib/sharedmount.php]: `$arguments['user']` is removed from the argument list of the method

**Comment by nickvergessen** [comment]
Running Smashbox

**Comment by nickvergessen** [comment]
Test | Result
---- | ------
./bin/smash lib/oc-tests/test_reshareDir.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_shareDir.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_shareFile.py -a | :x: 
./bin/smash lib/oc-tests/test_shareGroup.py -a | 
./bin/smash lib/oc-tests/test_sharePermissions.py -a | 
./bin/smash lib/oc-tests/test_uploadFiles.py -a | 
./bin/smash lib/test_basicSync.py -a | 


```
2015-06-30 11:21:09,457 - ERROR - sharer -
md5 of modified file /home/nickv/ownCloud/Smashbox/ramdir/test_shareFile-150630-112042-testset0/sharer/TEST_FILE_MODIFIED_USER_SHARE.dat did not change: expected 45b565af28a2998fa8d5dae7eb6e660e, got 45b565af28a2998fa8d5dae7eb6e660e
"md5 of modified file %s did not change: expected %s, got %s" % (fn, md5, actual_md5)) failed in expect_modified() ["/home/nickv/ownCloud/Smashbox/smashbox/python/smashbox/utilities/__init__.py" at line 684]
```

Exception in the owncloud.log:
```json
{
    "reqId": "IAyN/xfxYAeEZ1LyBFSq",
    "remoteAddr": "192.168.177.224",
    "app": "webdav",
    "message": {
        "Exception": {
            "Message": "HTTP/1.1 423 \"TEST_FILE_MODIFIED_USER_SHARE.dat\" is locked",
            "Code": 0,
            "Trace": {
                "#0": [
                    "/home/nickv/ownCloud/Smashbox/master/3rdparty/sabre/dav/lib/DAV/Server.php(1087): OC\Connector\Sabre\File->put(Resource id #22)",
                    null
                ],
                "#1": [
                    "/home/nickv/ownCloud/Smashbox/master/3rdparty/sabre/dav/lib/DAV/CorePlugin.php(501): Sabre\DAV\Server->updateFile('TEST_FILE_MODIF...', Resource id #22, NULL)",
                    null
                ],
                "#2": [
                    "[internal function]: Sabre\DAV\CorePlugin->httpPut(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))",
                    null
                ],
                "#3": [
                    "/home/nickv/ownCloud/Smashbox/master/3rdparty/sabre/event/lib/EventEmitterTrait.php(105): call_user_func_array(Array, Array)",
                    null
                ],
                "#4": [
                    "/home/nickv/ownCloud/Smashbox/master/3rdparty/sabre/dav/lib/DAV/Server.php(469): Sabre\Event\EventEmitter->emit('method:PUT', Array)",
                    null
                ],
                "#5": [
                    "/home/nickv/ownCloud/Smashbox/master/3rdparty/sabre/dav/lib/DAV/Server.php(254): Sabre\DAV\Server->invokeMethod(Object(Sabre\HTTP\Request), Object(Sabre\HTTP\Response))",
                    null
                ],
                "#6": [
                    "/home/nickv/ownCloud/Smashbox/master/apps/files/appinfo/remote.php(83):",
                    "Sabre\DAV\Server->exec()"
                ],
                "#7": [
                    "/home/nickv/ownCloud/Smashbox/master/remote.php(83):",
                    "require_once('/home/nickv/own...')"
                ],
                "#8": [
                    "{main}",
                    null
                ]
            },
            "File": "/home/nickv/ownCloud/Smashbox/master/lib/private/connector/sabre/file.php",
            "Line": 195
        }
    },
    "level": 4,
    "time": "2015-06-30T09:26:03+00:00"
}
```

File: apps/files_sharing/lib/sharedmount.php (hunk 1 above)
Comment (icewind1991) [inline apps/files_sharing/lib/sharedmount.php]: I think it's cleaner to pass it as argument

**Comment by icewind1991** [comment]
Fixes the failing smashbox test and added a unit test for it

**Comment by nickvergessen** [comment]
Test | Result
---- | ------
./bin/smash lib/oc-tests/test_reshareDir.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_shareDir.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_shareFile.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_shareGroup.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_sharePermissions.py -a | :white_check_mark: 
./bin/smash lib/oc-tests/test_uploadFiles.py -a | :white_check_mark: 
./bin/smash lib/test_basicSync.py -a | :white_check_mark:

**Comment by PVince81** [comment]
@nickvergessen does this nice table count as a thumbs up?

**Comment by PVince81** [comment]
Retested, still works :+1:

**Comment by PVince81** [comment]
Still needs a second reviewer

**Comment by nickvergessen** [comment]
Needs a rebase (and so a retest?)
Anyway will look in getting this merged then

**Comment by icewind1991** [comment]
rebased

**Comment by nickvergessen** [comment]
Tested with smashbox again, still passing:
:+1:

**Comment by nickvergessen** [comment]
Merging since the PR is against master.

@icewind1991 I guess you need to prepare a backport for 8.1.1 after 8.1.0 is out.

**Comment by MorrisJobke** [comment]
@karlitschek @icewind1991 Backport?

**Comment by karlitschek** [comment]
please backport

**Pull request merged**